sunpy/net/dataretriever/sources/gong.py

```diff
@@ -0,0 +1,282 @@
+    2016-07-04 00:00:00 2016-07-05 00:00:00   GONG Data not Available
+    2016-07-05 00:00:00 2016-07-06 00:00:00   GONG Data not Available]
+    """
+    def _get_url_for_timerange(self, timerange, **kwargs):
+        """
+        returns list of urls corresponding to given TimeRange.
+        """
+        #TO-DO: Figure out where scraper would and wouldn't work
+        table_physobs = {'INTENSITY' : 'i', 'LOS_MAGNETIC_FIELD': 'b'} #legitimate physical observations.
+        table_instruments = ['bb', 'ct', 'le', 'ml', 'td', 'ud','z'] #For magnetogram and intensity
+        
+        physobs_in = True if ('physobs' in kwargs.keys() and kwargs['physobs'] in table_physobs.keys()) else False #Is PhysObs entered
+        instrument_in = True if ('instrument' in kwargs.keys() and kwargs['instrument'] in table_instruments) else False
+        wavelength_in = True if 'wavelength' in kwargs.keys() else False
+        #Is instrument entered.
+        url_pattern_1 =  'ftp://gong2.nso.edu/QR/{id}qa/%Y%m/{ObsID}bqa%y%m%d/{ObsID}bqa%y%m%dt%H%M.fits.gz'
+        url_pattern_2 = 'http://gong2.nso.edu/HA/haf/%Y%m/%Y%m%d/%Y%m%d%H%M%S{ObsID}h.fits.fz'
+        
+        result = list() #final list of urls
+        patterns = list()
+        if not physobs_in:
+            patterns.append(url_pattern_1), patterns.append(url_pattern_2)
+        else:
+            if kwargs['physobs'] == 'LOS_MAGNETIC_FIELD':
+                patterns.append(url_pattern_1)
+            else:
+                #Differentiate on basis of wavelength
+                if not wavelength_in:
+                    patterns.append(url_pattern_1), patterns.append(url_pattern_2)
+                else:
+                    wave = int(kwargs['wavelength'].min.value)
+                    if (6562 <= wave <= 6563):
+                        patterns.append(url_pattern_2)
+                    elif wave == 6768:
+                        patterns.append(url_pattern_1)
+
+
+        #All valid patterns to be downloaded are in the patterns list.
+        instruments_to = list() #The instruments from which user wants to download
+        if not instrument_in:
+            instruments_to.extend(table_instruments)
+        else:
+            instruments_to.append(kwargs['instrument'])
+
+        
+        def download_from_nso(id, ObsID, time_range):
+            #Cannot scrape urls from NSO. The logic for downloading is
+            #dependent on NSO's policy of uploading data. Every day staring from 00:04
+            #fits files are uploaded regularly at an interval of 10 minutes.
+            result = list()
+            base_url = 'ftp://gong2.nso.edu/QR/{id}qa/{date:%Y%m}/{ObsID}{id}qa{date:%y%m%d}/{ObsID}{id}qa{date:%y%m%d}t{date:%H%M}.fits.gz'
+            total_days = (time_range.end - time_range.start).days + 1
+            all_dates = time_range.split(total_days)
+            for day in all_dates:
+                today = datetime.datetime(day.end.year, day.end.month, day.end.day,
```

Comment:
this is confusing, why are you using `day.end` ??  in fact... are you sure it's all right?
what happens if I query a time range like:     
```
Start: 2010-01-01 05:03:00
End:   2010-01-03 00:12:23
```
?????